src/ephemeris/set_library_permissions.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+'''Tool to set permissions for all datasets of a given Galaxy Data Library'''
+
+import argparse
+import logging as log
+from bioblend import galaxy
+import sys, time, os
+
+from .common_parser import get_common_args
+# Print iterations progress
+def printProgressBar (iteration, total, prefix = '', suffix = '', decimals = 1, length = 100, fill = '█', printEnd = "\r"):
+    """
+    Call in a loop to create terminal progress bar
+    @params:
+        iteration   - Required  : current iteration (Int)
+        total       - Required  : total iterations (Int)
+        prefix      - Optional  : prefix string (Str)
+        suffix      - Optional  : suffix string (Str)
+        decimals    - Optional  : positive number of decimals in percent complete (Int)
+        length      - Optional  : character length of bar (Int)
+        fill        - Optional  : bar fill character (Str)
+        printEnd    - Optional  : end character (e.g. "\r", "\r\n") (Str)
+    """
+    percent = ("{0:." + str(decimals) + "f}").format(100 * (iteration / float(total)))
+    filledLength = int(length * iteration // total)
+    bar = fill * filledLength + '-' * (length - filledLength)
+    print(f'\r{prefix} |{bar}| {percent}% {suffix}', end = printEnd)
+    # Print New Line on Complete
+    if iteration == total: 
+        print()
+
+def get_datasets(gi, library_id) -> [str]:
+    objects = gi.libraries.show_dataset(library_id=library_id, dataset_id='')
+    datasets = []
+    for index in range(len(objects)):
+       if objects[index]['type'] == 'file':
+           datasets.append(objects[index]['id'])
+    if datasets == []:
+        sys.exit("No datasets in library!")
+    else:
+        return datasets
+
+def set_permissions(gi, library_id, role_ids):
+    log.info("Your library_id is " + library_id + "\n")
+    role_msg = "Your roles are: " + " ".join(role_ids)
+    log.info(role_msg)
+    datasets = get_datasets(gi, library_id)
+    total = len(datasets)
+    est = total*3/60
+    # Give User time to abort
+    log.info('\nSuccess! %d datasets found. Processing can take up to %f min', total, est)
+    t = 5
+    while t:
+        log.info('Starting in %d s ... Press Crtl+C to abort.', t)
+        time.sleep(1)
+        t -= 1
```

Comment:
```suggestion
```

I don't know if this is necessary? If you want to do this, we should default to "dry run" and then have a `--commit` flag that would actually do the change permissions call

Reply:
I have this because the API is quite slow in this case and it took ~50 min to change 940 datasets, so I would like to inform the user about the estimated time and give the option to abort before it's starting

Reply:
Yes, hence doing the converse:

- doing a dry run by default (which can give a time estimate0
- and then running if they request it

Rather than having a short window in which, if they don't read fast enough, things can start happening.

The other "safe" alternative is doing an `input()` which will block on user input (so you can't accidentally have it run by not reading fast enough about what's happening). And checking if they say `y` to continue.

Reply:
I added the y/n question

Reply:
> I have this because the API is quite slow in this case and it took ~50 min to change 940 datasets, so I would like to inform the user about the estimated time and give the option to abort before it's starting

Can you create an issue in Galaxy for this ? There's no intrinsic reason that this needs to be so slow.